File: leetcode/array/3437_Permutations_III/codes/backtrack.py
from typing import List

class Solution:
    def permute(self, n: int) -> List[List[int]]:
        res = []
        used = [False] * (n + 1)
        def backtrack(path):
            if len(path) == n:
                res.append(path[:])
                return
            for num in range(1, n + 1):
                if not used[num]:
                    if not path or (path[-1] % 2 != num % 2):
                        used[num] = True
                        path.append(num)
                        backtrack(path)
                        path.pop()
                        used[num] = False
        backtrack([])
        return res

File: leetcode/array/3437_Permutations_III/codes/backtrack.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    vector<vector<int>> permute(int n) {
        vector<vector<int>> res;
        vector<int> path;
        vector<bool> used(n + 1, false);
        function<void()> backtrack = [&]() {
            if (path.size() == n) {
                res.push_back(path);
                return;
            }
            for (int num = 1; num <= n; num++) {
                if (!used[num]) {
                    if (path.empty() || (path.back() % 2 != num % 2)) {
                        used[num] = true;
                        path.push_back(num);
                        backtrack();
                        path.pop_back();
                        used[num] = false;
                    }
                }
            }
        };
        backtrack();
        return res;
    }
};
